services/ClusterService_test.go

```diff
@@ -0,0 +1,1162 @@
+						gomega.Expect(err).To(gomega.HaveOccurred())
+					})
+				})
+			})
+		})
+
+		ginkgo.Context("when cluster doesn't exist", func() {
+			ginkgo.BeforeEach(func() {
+				recName = clusterUpdateUnknown
+
+				cluster = resources.CreateClusterWithID(nonExistingClusterID)
+				if cluster == nil {
+					err = errors.ErrNoCluster
+				}
+
+				clusterBlueprint = blueprint.CreateUpdateClusterBlueprint()
+				if clusterBlueprint == nil {
+					err = errors.ErrNoClusterBlueprint
+					return
+				}
+				clusterBlueprint.SetReservedCPU("dummy")
+			})
+
+			ginkgo.It("should return that cluster with given ID doesn't exist", func() {
+				gomega.Expect(err).NotTo(gomega.HaveOccurred()) // no error during BeforeEach
+
+				err = client.ClusterService.Update(context.TODO(), *cluster, clusterBlueprint, services.Merge)
+				gomega.Expect(err).To(gomega.HaveOccurred())
+			})
+		})
+
+		ginkgo.Context("when cluster is empty", func() {
+			ginkgo.BeforeEach(func() {
+				recName = clusterUpdateNoUser
+
+				clusterBlueprint = blueprint.CreateUpdateClusterBlueprint()
+				if clusterBlueprint == nil {
+					err = errors.ErrNoClusterBlueprint
+					return
+				}
+				clusterBlueprint.SetReservedCPU("dummy")
+			})
+
+			ginkgo.It("should return that cluster has no ID", func() {
+				gomega.Expect(err).NotTo(gomega.HaveOccurred()) // no error during BeforeEach
+
+				err = client.ClusterService.Update(context.TODO(), resources.Cluster{},
+					clusterBlueprint, services.Merge)
+				gomega.Expect(err).To(gomega.HaveOccurred())
+			})
+		})
+	})
+
+	ginkgo.Describe("cluster add host", func() {
+		var (
+			cluster           *resources.Cluster
+			oneCluster        *resources.Cluster
+			clusterID         int
+			host              *resources.Host
+			existingHostID    = 4
+			nonExistingHostID = 110
+		)
+
+		ginkgo.Context("when cluster exists", func() {
+			ginkgo.BeforeEach(func() {
+				cluster = resources.CreateClusterWithID(existingClusterID)
+				if cluster == nil {
+					err = errors.ErrNoCluster
+				}
+			})
+
+			ginkgo.When("when new host exists", func() {
+				ginkgo.BeforeEach(func() {
+					recName = clusterAddHost
+
+					host = resources.CreateHostWithID(existingHostID)
+				})
+
+				ginkgo.It("should add host to the given cluster", func() {
+					gomega.Expect(err).NotTo(gomega.HaveOccurred()) // no error during BeforeEach
+
+					err = client.ClusterService.AddHost(context.TODO(), *cluster, *host)
+					gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+					// check whether host was really added in OpenNebula
+					clusterID, err = cluster.ID()
+					gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+					oneCluster, err = client.ClusterService.RetrieveInfo(context.TODO(), clusterID)
+					gomega.Expect(err).NotTo(gomega.HaveOccurred())
+					gomega.Expect(oneCluster).ShouldNot(gomega.BeNil())
+
+					var hosts []int
+					hosts, err = oneCluster.Hosts()
+					gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
+					founded := false
```

Comment:
When checking whether some element can be found within an array use specialized matcher for that:
[https://onsi.github.io/gomega/#containelementelement-interface](https://onsi.github.io/gomega/#containelementelement-interface)